### DISCUSSION TITLE
Prediction methods for multi-topic documents

### DISCUSSION THREAD

**Discussion by ghost** [MaartenGr/BERTopic#1051] [Q&A] [answered]
Hello, 

I just trained a bertopic model on customer reviews that are short and of poor quality (about 10/15 words). The performances obtained are very satisfactory ! On the other hand, the goal of my project is to make multi-label classification what BERTopic does not manage. So I am looking for the best solution to get around this problem.

I want to predict the distribution of my document (passed in input) to all the topics identified by my model, based on my trained model. 
To do this, I have identified several methods but I would like to know the advantages and disadvantages and what are the best practices. 

- `.transform` allows to classify a new document when the model is trained. By default, we can only get the probability of belonging to the subject that has been identified. In other words, we cannot obtain the probability of belonging to the other subjects. If you want to do this, then you should activate the parameter (calculate_probabilities) when training the model. **I wonder how the probability is calculated (technical details of the prediction process)**

- `approximate_distribution` this method allows to compare the groups of tokens of a document specified in input with each of the subjects of the model and returns the distribution of membership of the document with each of these topics. This is the method that seems most appropriate, in my opinion, for the kind of task I have to perform.

- `find_topics` this method converts a string of words passed in parameter into embeddings and will compare this embedding with all the embeddings of each topic in the model (you have to be careful not to pass too many words in input by maybe applying some pre-processing on the customer review in input). 

I would like to know more about the process applied when calling the `.transform` function? What is the difference between `.transform` and `.find_topics` ? The (mis)understanding I have is that `.transform` applies the `.find_topics` on the whole document. 

Is it possible to further explain the difference between these methods and which one would be more appropriate in a multi-label classification project.

Thank you very much in advance,
Thomas

**Comment by MaartenGr** [comment, marked as answer]
Glad to hear that you are happy with the preliminary results! 

> .transform allows to classify a new document when the model is trained. By default, we can only get the probability of belonging to the subject that has been identified. In other words, we cannot obtain the probability of belonging to the other subjects. If you want to do this, then you should activate the parameter (calculate_probabilities) when training the model. I wonder how the probability is calculated (technical details of the prediction process)

That probability is generated as a result of HDBSCAN's soft-clustering process. Without going in too much detail, it creates an approximation of soft-assignment after the cluster model was fitted. You can find more about that [here](https://hdbscan.readthedocs.io/en/latest/soft_clustering.html) and [here](https://hdbscan.readthedocs.io/en/latest/how_hdbscan_works.html). 

> approximate_distribution this method allows to compare the groups of tokens of a document specified in input with each of the subjects of the model and returns the distribution of membership of the document with each of these topics. This is the method that seems most appropriate, in my opinion, for the kind of task I have to perform.

Yes, based on your description, this seems like it would suit your use case best as it allows for topic distributions to be created. More specifically, it applies a sliding window over pieces of text in the document to classify the distribution of topics using either c-TF-IDF or embedding similarity. 

> I would like to know more about the process applied when calling the .transform function? What is the difference between .transform and .find_topics ? The (mis)understanding I have is that .transform applies the .find_topics on the whole document.

`.find_topics` is a different method from `.transform` in that it quickly tries to find a topic that is similar to an input word. The underlying process between those two methods are quite different from one another. `.transform` applies the pipeline BERTopic is known for (embeddings -> dimensionality reduction -> clustering) and essentially does its prediction with the clustering algorithm. This is the most accurate method of making predictions as it follows the exact same procedure as `.fit`. In contrast, `.find_topics` merely calculates the embeddings of its inputs and compares that with the topic embeddings. The most similar topic embeddings are then selected. This is a bit more rough and is typically not intended for finding which topic a document belongs to. The main use case for this method is wanting to find out which of the topics are about `"cars"` by simply providing that keyword.

**Reply by ghost** [reply]
Okay perfect, thank you for your answer which is very clear. However, it seems to me that the two links mentioned do not work.

**Reply by MaartenGr** [reply]
Oops, forgot to add the links, I edited the message and it should contain the correct links now.
